### ISSUE TITLE
Pyright confuses union type with and type

### ISSUE THREAD

**Issue by jlricon** [microsoft/pyright#3894]
**Describe the bug**
The code below should pass as Dict[str,str] is a narrower type than Dict[str,str|None] but it doesn't

**To Reproduce**
Typecheck the code below

**Expected behavior**

**Screenshots or Code**
In the code below, pyright complains about passing test_dict to test, saying that
```
Argument of type "dict[str, str]" cannot be assigned to parameter "types" of type "dict[str, str | None]" in function "test"
  TypeVar "_VT@dict" is invariant
    Type "str" cannot be assigned to type "str | None"
      Type cannot be assigned to type "None"
```
```python

def test(types: dict[str, str | None]):
    pass


test_dict = {"a": "b"}
test(test_dict)
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Command line, version 1.1.269

**Comment by erictraut** [comment]
Pyright is correct here. The type parameters for `dict` are invariant, so `dict[str, str]` is not a subtype of `dict[str, str | None]`. Mypy gives the same error in this case, as is expected.

For additional details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#generic-types).
